scanner/datastore/postgres/external_index_report.go

```diff
@@ -0,0 +1,142 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/pkg/errors"
+	"github.com/quay/claircore"
+	"github.com/quay/zlog"
+)
+
+// ErrDidNotUpdate indicates the row was not updated.
+var ErrDidNotUpdate = errors.New("")
+
+const selectIndexReport = `
+	SELECT indexer_version FROM external_index_report
+	WHERE hash_id = $1 FOR UPDATE`
+
+// StoreIndexReport stores an external index report, indexReport, with an ID
+// based on hashID with expiration. On hashID conflict, StoreIndexReport will
+// use versionCmpFn to determine whether to update the row during the
+// transaction. If hashID does not currently exist in the external index report
+// datastore, it will be created.
+func (e *externalIndexStore) StoreIndexReport(
+	ctx context.Context,
+	hashID string,
+	indexerVersion string,
+	indexReport *claircore.IndexReport,
+	expiration time.Time,
+	versionCmpFn func(iv string) bool,
+) error {
+	ctx = zlog.ContextWithValues(
+		ctx,
+		"component",
+		"datastore/postgres/externalIndexStore.StoreIndexReport",
+	)
+
+	var storedIndexerVersion string
+	err := e.pool.QueryRow(ctx, selectIndexReport, hashID).Scan(&storedIndexerVersion)
+	if err != nil {
+		// If no rows were found, create the external index report.
+		if errors.Is(err, pgx.ErrNoRows) {
+			const insertIndexReport = `
+				INSERT INTO external_index_report (hash_id, indexer_version, index_report, expiration) VALUES
+					($1, $2, $3, $4)`
+			_, err := e.pool.Exec(
+				ctx,
+				insertIndexReport,
+				hashID,
+				indexerVersion,
+				indexReport,
+				expiration.UTC(),
+			)
+			if err != nil {
+				return fmt.Errorf("storing new external index report: %w", err)
+			}
+			return nil
+		}
+
+		return fmt.Errorf("querying external index reports: %w", err)
+	}
+
+	if !versionCmpFn(storedIndexerVersion) {
+		return fmt.Errorf("stored index report was produced with more recent indexer: %w", ErrDidNotUpdate)
+	}
+
+	const updateIndexReport = `
+		UPDATE external_index_report SET (indexer_version, index_report, expiration) =
+			($2, $3, $4) WHERE hash_id = $1`
+
+	_, err = e.pool.Exec(ctx, updateIndexReport, hashID, indexerVersion, indexReport, expiration)
+	if err != nil {
+		return fmt.Errorf("updating index report: %w", err)
+	}
+
+	return nil
+}
+
+func (e *externalIndexStore) GCIndexReports(
+	ctx context.Context,
+	expiration time.Time,
+	opts ...ReindexGCOption,
+) ([]string, error) {
+	o := makeReindexGCOpts(opts)
+
+	ctx = zlog.ContextWithValues(
+		ctx,
+		"component",
+		"datastore/postgres/externalIndexStore.GCIndexReports",
+	)
+
+	const deleteIndexReports = `
+		DELETE FROM external_index_report
+		WHERE hash_id IN (
+		    SELECT hash_id FROM external_index_report WHERE expiration < $1 LIMIT $2
+		)
+		RETURNING hash_id`
```

Comment:
Can this be simplified?
```suggestion
		DELETE FROM external_index_report 
		WHERE expiration < $1 
		LIMIT $2
		RETURNING hash_id`
```

Reply:
TIL